File: modules/checks/disclosure.py
from __future__ import annotations
import re
from urllib.parse import urljoin
import requests

from ..findings import Finding, Severity
from ..session import session_to_curl_flags


SENSITIVE_PATHS = {
    # API docs
    "/swagger.json": "Swagger JSON (API schema)",
    "/swagger.yaml": "Swagger YAML",
    "/swagger-ui.html": "Swagger UI",
    "/swagger-ui/": "Swagger UI",
    "/api-docs": "API docs",
    "/openapi.json": "OpenAPI schema",
    "/openapi.yaml": "OpenAPI schema",
    "/graphql": "GraphQL endpoint",
    "/graphiql": "GraphiQL IDE",
    "/v1/api-docs": "Spring API docs",
    "/v2/api-docs": "Spring API docs v2",
    # Actuator / management
    "/actuator": "Spring Boot Actuator",
    "/actuator/env": "Spring Boot env (env vars!)",
    "/actuator/health": "Spring Boot health",
    "/actuator/metrics": "Spring Boot metrics",
    "/actuator/beans": "Spring Boot beans",
    "/actuator/mappings": "Spring Boot mappings",
    "/actuator/trace": "Spring Boot request trace",
    "/actuator/httptrace": "Spring Boot HTTP trace",
    "/actuator/logfile": "Spring Boot logfile",
    "/actuator/threaddump": "Spring Boot thread dump",
    "/actuator/heapdump": "Spring Boot heap dump!",
    "/actuator/sessions": "Spring Boot sessions",
    "/actuator/scheduledtasks": "Spring Boot scheduled tasks",
    # Debug
    "/debug": "debug endpoint",
    "/console": "debug console",
    "/phpinfo.php": "PHP info",
    "/info.php": "PHP info",
    "/test.php": "PHP test file",
    "/.env": ".env file",
    "/.env.local": ".env.local",
    "/.env.development": ".env.development",
    "/.env.production": ".env.production",
    # Config / Git
    "/.git/config": "Git config",
    "/.git/HEAD": "Git HEAD",
    "/.svn/entries": "SVN entries",
    "/web.config": "IIS web.config",
    "/.htaccess": "Apache .htaccess",
    "/config.json": "config.json",
    "/config.yaml": "config.yaml",
    "/settings.json": "settings.json",
    "/app.config": "app.config",
    # Backup
    "/backup.zip": "backup archive",
    "/backup.sql": "SQL dump",
    "/db.sql": "SQL dump",
    "/dump.sql": "SQL dump",
    # Secrets
    "/credentials.json": "credentials",
    "/secrets.json": "secrets",
    "/private.key": "private key",
    "/server.key": "SSL private key",
    # Metrics / monitoring
    "/metrics": "metrics endpoint",
    "/prometheus": "Prometheus metrics",
    "/health": "health check",
    "/status": "status endpoint",
    "/ping": "ping",
    # Admin
    "/admin": "admin panel",
    "/admin/login": "admin login",
    "/administrator": "administrator panel",
    "/manage": "management panel",
    "/wp-admin": "WordPress admin",
    "/phpmyadmin": "phpMyAdmin",
}

ERROR_TRIGGERS = [
    "/'\"<>{}[]()", # various injection chars
    "../../../../../etc/passwd",
    "undefined",
    "\x00",
    "' OR 1=1--",
]

# Patterns indicating verbose errors
VERBOSE_ERROR_SIGNATURES = [
    r"Traceback \(most recent call last\)",      # Python
    r"at .+\([\w.]+:\d+\)",                      # Java stack trace
    r"NullPointerException",
    r"ClassNotFoundException",
    r"Exception in thread",
    r"PHP Fatal error",
    r"PHP Warning",
    r"PHP Notice",
    r"Parse error:.*PHP",
    r"Warning:.*on line \d+",
    r"include\(.*\): failed to open stream",
    r"Microsoft.*ODBC.*Driver",
    r"ORA-\d{5}:",
    r"SQLSTATE\[",
    r"You have an error in your SQL syntax",
    r"Uncaught Error:",
    r"Unhandled Promise Rejection",
    r"undefined is not a function",             # JS errors
    r"Cannot read property .* of undefined",
    r"Error: ENOENT: no such file",             # Node.js
    r"ActiveRecord::.*Error",                   # Rails
    r"PG::.*Error",
    r"django\.core\.exceptions",
    r"WSGI Application Error",
]

VERSION_PATTERNS = {
    "Apache": re.compile(r"Apache/([\d.]+)", re.I),
    "nginx": re.compile(r"nginx/([\d.]+)", re.I),
    "PHP": re.compile(r"PHP/([\d.]+)", re.I),
    "ASP.NET": re.compile(r"ASP\.NET Version:([\d.]+)", re.I),
    "Spring": re.compile(r"Spring Framework ([\d.]+)", re.I),
    "Tomcat": re.compile(r"Apache Tomcat/([\d.]+)", re.I),
    "Express": re.compile(r"Express/([\d.]+)", re.I),
    "Django": re.compile(r"Django/([\d.]+)", re.I),
    "Rails": re.compile(r"Rails ([\d.]+)", re.I),
}

SECRET_PATTERNS = {
    "AWS Access Key": re.compile(r"AKIA[0-9A-Z]{16}"),
    "AWS Secret Key": re.compile(r"(?i)aws_secret_access_key\s*[=:]\s*([A-Za-z0-9/+]{40})"),
    "Private Key": re.compile(r"-----BEGIN (RSA |EC |DSA )?PRIVATE KEY-----"),
    "Google API Key": re.compile(r"AIza[0-9A-Za-z\-_]{35}"),
    "Stripe Key": re.compile(r"sk_(live|test)_[0-9a-zA-Z]{24,}"),
    "GitHub Token": re.compile(r"gh[pousr]_[A-Za-z0-9_]{36,255}"),
    "JWT Secret": re.compile(r"(?i)(jwt[_-]?secret|secret[_-]?key)\s*[=:]\s*['\"]([^'\"]{8,})"),
    "Database URL": re.compile(r"(?i)(postgres|mysql|mongodb|redis)://[^@\s]{4,}@[^\s\"']+"),
}


def run(session: requests.Session, base_url: str, endpoints: list, store) -> None:
    print("[*] Checking information disclosure...")
    curl_auth = session_to_curl_flags(session)
    timeout = getattr(session, 'timeout', 15)

    _check_sensitive_paths(session, base_url, curl_auth, timeout, store)
    _check_verbose_errors(session, base_url, endpoints, curl_auth, timeout, store)
    _check_secrets_in_responses(session, endpoints, curl_auth, timeout, store)
    _check_graphql(session, base_url, curl_auth, timeout, store)


def _check_sensitive_paths(session, base_url, curl_auth, timeout, store):
    for path, label in SENSITIVE_PATHS.items():
        url = base_url.rstrip("/") + path
        try:
            resp = session.get(url, timeout=8, allow_redirects=True)
        except Exception:
            continue

        if resp.status_code in (404, 410, 501):
            continue

        severity = Severity.HIGH
        if resp.status_code in (401, 403):
            severity = Severity.INFO
        elif path in ("/.env", "/.git/config", "/actuator/env", "/actuator/heapdump",
                      "/backup.sql", "/credentials.json", "/secrets.json", "/private.key"):
            severity = Severity.CRITICAL

        ct = resp.headers.get("content-type", "").lower()
        body_snippet = resp.text[:300].replace("\n", " ")

        store.add(Finding(
            title=f"Доступен чувствительный путь: {path} ({label})",
            severity=severity,
            category="Information Disclosure",
            cwe="CWE-200",
            description=(
                f"URL '{url}' возвращает HTTP {resp.status_code}.\n"
                f"Ресурс: {label}\n"
                f"Content-Type: {ct or 'не указан'}"
            ),
            url=url,
            evidence=f"HTTP {resp.status_code}\n{body_snippet}",
            remediation=(
                f"1. Закройте доступ к {path} на уровне веб-сервера / файрвола.\n"
                "2. Удалите чувствительные файлы из webroot.\n"
                "3. Ограничьте Actuator-эндпоинты (management.endpoints.web.exposure.include)."
            ),
            reproduction=f"curl -sk {curl_auth} '{url}'",
        ))


def _check_verbose_errors(session, base_url, endpoints, curl_auth, timeout, store):
    tested = set()
    targets = [base_url] + [ep.url for ep in endpoints if ep.method == "GET"][:20]

    for url in targets:
        if url in tested:
            continue
        tested.add(url)

        for trigger in ERROR_TRIGGERS[:3]:
            test_url = url + trigger if "?" not in url else url + "&err=" + trigger
            try:
                resp = session.get(test_url, timeout=8)
            except Exception:
                continue

            for sig in VERBOSE_ERROR_SIGNATURES:
                m = re.search(sig, resp.text, re.IGNORECASE)
                if m:
                    snippet = resp.text[max(0, m.start()-100):m.end()+200]
                    store.add(Finding(
                        title="Verbose error messages — утечка стектрейса",
                        severity=Severity.MEDIUM,
                        category="Information Disclosure",
                        cwe="CWE-209",
                        description=(
                            f"Приложение возвращает подробные сообщения об ошибках.\n"
                            f"Паттерн: {sig}\n"
                            "Это раскрывает технологический стек, пути файлов и детали реализации."
                        ),
                        url=test_url,
                        evidence=f"...{snippet}...",
                        remediation=(
                            "1. Настройте production error handler: возвращайте только generic сообщение.\n"
                            "2. Логируйте детали ошибки только на сервере.\n"
                            "3. DEBUG=False (Django), displayErrors: false, error_reporting(0) (PHP)."
                        ),
                        reproduction=f"curl -sk {curl_auth} '{test_url}'",
                    ))
                    tested.add(url)  # Don't re-test this URL
                    break
            break  # One trigger per URL is enough for detection


def _check_secrets_in_responses(session, endpoints, curl_auth, timeout, store):
    for ep in endpoints[:30]:  # Limit to first 30 endpoints
        if ep.method not in ("GET", "HEAD"):
            continue
        try:
            resp = session.get(ep.url, timeout=timeout)
            if resp.status_code != 200:
                continue
            text = resp.text
        except Exception:
            continue

        for secret_type, pattern in SECRET_PATTERNS.items():
            m = pattern.search(text)
            if m:
                store.add(Finding(
                    title=f"Секрет в HTTP-ответе: {secret_type}",
                    severity=Severity.CRITICAL,
                    category="Information Disclosure",
                    cwe="CWE-312",
                    description=(
                        f"В ответе '{ep.url}' обнаружен {secret_type}.\n"
                        "Утечка credentials в HTTP-ответах критична для безопасности."
                    ),
                    url=ep.url,
                    evidence=f"Найдено: {m.group()[:80]}...",
                    remediation=(
                        "1. Немедленно ротируйте скомпрометированные credentials.\n"
                        "2. Не возвращайте секреты в API-ответах.\n"
                        "3. Используйте secret management (Vault, AWS Secrets Manager).\n"
                        "4. Проверьте git-историю на предмет закоммиченных секретов."
                    ),
                    reproduction=f"curl -sk {curl_auth} '{ep.url}' | grep -oE '<pattern>'",
                ))


def _check_graphql(session, base_url, curl_auth, timeout, store):
    graphql_endpoints = ["/graphql", "/api/graphql", "/graphiql", "/gql", "/query"]

    for path in graphql_endpoints:
        url = base_url.rstrip("/") + path
        # Check if endpoint exists
        try:
            probe = session.get(url, timeout=8)
            if probe.status_code in (404, 410):
                continue
        except Exception:
            continue

        # Try introspection
        introspection_query = {
            "query": "{ __schema { types { name fields { name } } } }"
        }
        try:
            resp = session.post(
                url,
                json=introspection_query,
                headers={"Content-Type": "application/json"},
                timeout=timeout,
            )
            if resp.status_code == 200 and "__schema" in resp.text:
                store.add(Finding(
                    title="GraphQL: introspection включена (schema disclosure)",
                    severity=Severity.MEDIUM,
                    category="Information Disclosure",
                    cwe="CWE-200",
                    description=(
                        f"GraphQL endpoint '{url}' возвращает полную схему через introspection.\n"
                        "Атакующий может восстановить всю структуру API, включая мутации и аргументы."
                    ),
                    url=url,
                    evidence=f"__schema присутствует в ответе на introspection-запрос",
                    remediation=(
                        "1. Отключите introspection в production:\n"
                        "   GraphQL Shield, depth-limit или параметр disableIntrospection.\n"
                        "2. Если introspection нужна — ограничьте по IP или роли."
                    ),
                    reproduction=(
                        f"curl -sk {curl_auth} -X POST '{url}' \\\n"
                        f"  -H 'Content-Type: application/json' \\\n"
                        f"  -d '{{\"query\":\"{{ __schema {{ types {{ name }} }} }}\"}}"
                    ),
                ))

            # Check for batch query attack
            batch_query = [introspection_query, introspection_query, introspection_query]
            resp_batch = session.post(
                url,
                json=batch_query,
                headers={"Content-Type": "application/json"},
                timeout=timeout,
            )
            if resp_batch.status_code == 200 and isinstance(resp_batch.json(), list):
                store.add(Finding(
                    title="GraphQL: batch queries включены",
                    severity=Severity.LOW,
                    category="API Security",
                    cwe="CWE-770",
                    description=(
                        "GraphQL endpoint принимает batch-запросы (массив операций в одном HTTP-запросе). "
                        "Это может использоваться для обхода rate limiting или брутфорса."
                    ),
                    url=url,
                    evidence=f"Batch ответ: {resp_batch.text[:200]}",
                    remediation="Ограничьте глубину и количество операций в одном запросе. Отключите batch если не используется.",
                    reproduction=(
                        f"curl -sk {curl_auth} -X POST '{url}' \\\n"
                        f"  -H 'Content-Type: application/json' \\\n"
                        f"  -d '[{{\"query\":\"{{me{{id}}}}\"}},{{\"query\":\"{{me{{id}}}}\"}},{{\"query\":\"{{me{{id}}}}\"}}]'"
                    ),
                ))
        except Exception:
            pass

File: go-vectrix/internal/checks/disclosure.go
package checks

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
	"time"

	"vectrixgo/internal/crawler"
	"vectrixgo/internal/findings"
	"vectrixgo/internal/httpsession"
	"vectrixgo/internal/logging"
)

type sensitivePath struct {
	path  string
	label string
}

// sensitivePaths is checked against base_url. Order mirrors
// modules/checks/disclosure.py SENSITIVE_PATHS (dict insertion order).
var sensitivePaths = []sensitivePath{
	{"/swagger.json", "Swagger JSON (API schema)"},
	{"/swagger.yaml", "Swagger YAML"},
	{"/swagger-ui.html", "Swagger UI"},
	{"/swagger-ui/", "Swagger UI"},
	{"/api-docs", "API docs"},
	{"/openapi.json", "OpenAPI schema"},
	{"/openapi.yaml", "OpenAPI schema"},
	{"/graphql", "GraphQL endpoint"},
	{"/graphiql", "GraphiQL IDE"},
	{"/v1/api-docs", "Spring API docs"},
	{"/v2/api-docs", "Spring API docs v2"},
	{"/actuator", "Spring Boot Actuator"},
	{"/actuator/env", "Spring Boot env (env vars!)"},
	{"/actuator/health", "Spring Boot health"},
	{"/actuator/metrics", "Spring Boot metrics"},
	{"/actuator/beans", "Spring Boot beans"},
	{"/actuator/mappings", "Spring Boot mappings"},
	{"/actuator/trace", "Spring Boot request trace"},
	{"/actuator/httptrace", "Spring Boot HTTP trace"},
	{"/actuator/logfile", "Spring Boot logfile"},
	{"/actuator/threaddump", "Spring Boot thread dump"},
	{"/actuator/heapdump", "Spring Boot heap dump!"},
	{"/actuator/sessions", "Spring Boot sessions"},
	{"/actuator/scheduledtasks", "Spring Boot scheduled tasks"},
	{"/debug", "debug endpoint"},
	{"/console", "debug console"},
	{"/phpinfo.php", "PHP info"},
	{"/info.php", "PHP info"},
	{"/test.php", "PHP test file"},
	{"/.env", ".env file"},
	{"/.env.local", ".env.local"},
	{"/.env.development", ".env.development"},
	{"/.env.production", ".env.production"},
	{"/.git/config", "Git config"},
	{"/.git/HEAD", "Git HEAD"},
	{"/.svn/entries", "SVN entries"},
	{"/web.config", "IIS web.config"},
	{"/.htaccess", "Apache .htaccess"},
	{"/config.json", "config.json"},
	{"/config.yaml", "config.yaml"},
	{"/settings.json", "settings.json"},
	{"/app.config", "app.config"},
	{"/backup.zip", "backup archive"},
	{"/backup.sql", "SQL dump"},
	{"/db.sql", "SQL dump"},
	{"/dump.sql", "SQL dump"},
	{"/credentials.json", "credentials"},
	{"/secrets.json", "secrets"},
	{"/private.key", "private key"},
	{"/server.key", "SSL private key"},
	{"/metrics", "metrics endpoint"},
	{"/prometheus", "Prometheus metrics"},
	{"/health", "health check"},
	{"/status", "status endpoint"},
	{"/ping", "ping"},
	{"/admin", "admin panel"},
	{"/admin/login", "admin login"},
	{"/administrator", "administrator panel"},
	{"/manage", "management panel"},
	{"/wp-admin", "WordPress admin"},
	{"/phpmyadmin", "phpMyAdmin"},
}

// errorTriggers — only the first 3 are used (mirrors ERROR_TRIGGERS[:3]).
var errorTriggers = []string{
	`/'"<>{}[]()`,
	"../../../../../etc/passwd",
	"undefined",
	"\x00",
	"' OR 1=1--",
}

var verboseErrorSignatures = []*regexp.Regexp{
	regexp.MustCompile(`(?i)Traceback \(most recent call last\)`),
	regexp.MustCompile(`(?i)at .+\([\w.]+:\d+\)`),
	regexp.MustCompile(`(?i)NullPointerException`),
	regexp.MustCompile(`(?i)ClassNotFoundException`),
	regexp.MustCompile(`(?i)Exception in thread`),
	regexp.MustCompile(`(?i)PHP Fatal error`),
	regexp.MustCompile(`(?i)PHP Warning`),
	regexp.MustCompile(`(?i)PHP Notice`),
	regexp.MustCompile(`(?i)Parse error:.*PHP`),
	regexp.MustCompile(`(?i)Warning:.*on line \d+`),
	regexp.MustCompile(`(?i)include\(.*\): failed to open stream`),
	regexp.MustCompile(`(?i)Microsoft.*ODBC.*Driver`),
	regexp.MustCompile(`(?i)ORA-\d{5}:`),
	regexp.MustCompile(`(?i)SQLSTATE\[`),
	regexp.MustCompile(`(?i)You have an error in your SQL syntax`),
	regexp.MustCompile(`(?i)Uncaught Error:`),
	regexp.MustCompile(`(?i)Unhandled Promise Rejection`),
	regexp.MustCompile(`(?i)undefined is not a function`),
	regexp.MustCompile(`(?i)Cannot read property .* of undefined`),
	regexp.MustCompile(`(?i)Error: ENOENT: no such file`),
	regexp.MustCompile(`(?i)ActiveRecord::.*Error`),
	regexp.MustCompile(`(?i)PG::.*Error`),
	regexp.MustCompile(`(?i)django\.core\.exceptions`),
	regexp.MustCompile(`(?i)WSGI Application Error`),
}

type secretPattern struct {
	name string
	re   *regexp.Regexp
}

var secretPatterns = []secretPattern{
	{"AWS Access Key", regexp.MustCompile(`AKIA[0-9A-Z]{16}`)},
	{"AWS Secret Key", regexp.MustCompile(`(?i)aws_secret_access_key\s*[=:]\s*([A-Za-z0-9/+]{40})`)},
	{"Private Key", regexp.MustCompile(`-----BEGIN (RSA |EC |DSA )?PRIVATE KEY-----`)},
	{"Google API Key", regexp.MustCompile(`AIza[0-9A-Za-z\-_]{35}`)},
	{"Stripe Key", regexp.MustCompile(`sk_(live|test)_[0-9a-zA-Z]{24,}`)},
	{"GitHub Token", regexp.MustCompile(`gh[pousr]_[A-Za-z0-9_]{36,255}`)},
	{"JWT Secret", regexp.MustCompile(`(?i)(jwt[_-]?secret|secret[_-]?key)\s*[=:]\s*['"]([^'"]{8,})`)},
	{"Database URL", regexp.MustCompile(`(?i)(postgres|mysql|mongodb|redis)://[^@\s]{4,}@[^\s"']+`)},
}

// RunDisclosure checks for exposed sensitive paths, verbose error messages,
// secrets leaked in responses, and GraphQL introspection/batch issues.
// Mirrors modules/checks/disclosure.py run().
func RunDisclosure(session *httpsession.Session, baseURL string, endpoints []crawler.Endpoint, store *findings.FindingStore) {
	logging.Println("[*] Checking information disclosure...")
	curlAuth := session.CurlAuthFlags(baseURL)

	checkSensitivePaths(session, baseURL, curlAuth, store)
	checkVerboseErrors(session, baseURL, endpoints, curlAuth, store)
	checkSecretsInResponses(session, endpoints, store)
	checkGraphQL(session, baseURL, curlAuth, store)
}

type fallbackBaseline struct {
	status int
	body   string
}

// getFallbackBaseline requests a guaranteed-nonexistent path so SPA catch-all
// routing (200 + identical body for any path) can be filtered out.
func getFallbackBaseline(session *httpsession.Session, baseURL string) *fallbackBaseline {
	b := make([]byte, 16)
	rand.Read(b)
	probePath := "/__nonexistent_" + hex.EncodeToString(b) + "__"
	u := strings.TrimRight(baseURL, "/") + probePath
	resp, err := session.Request("GET", u, httpsession.Options{AllowRedirects: true, Timeout: 8 * time.Second})
	if err != nil {
		return nil
	}
	return &fallbackBaseline{status: resp.StatusCode, body: resp.Body}
}

func looksLikeFallback(resp *httpsession.Response, baseline *fallbackBaseline) bool {
	if baseline == nil {
		return false
	}
	return resp.StatusCode == baseline.status && resp.Body == baseline.body
}

func checkSensitivePaths(session *httpsession.Session, baseURL, curlAuth string, store *findings.FindingStore) {
	baseline := getFallbackBaseline(session, baseURL)

	for _, sp := range sensitivePaths {
		u := strings.TrimRight(baseURL, "/") + sp.path
		resp, err := session.Request("GET", u, httpsession.Options{AllowRedirects: true, Timeout: 8 * time.Second})
		if err != nil {
			continue
		}

		if resp.StatusCode == 404 || resp.StatusCode == 410 || resp.StatusCode == 501 {
			continue
		}
		if looksLikeFallback(resp, baseline) {
			continue
		}

		ct := strings.ToLower(resp.Header.Get("Content-Type"))
		bodySnippet := strings.ReplaceAll(truncate(resp.Body, 300), "\n", " ")

		if resp.StatusCode == 401 || resp.StatusCode == 403 {
			f := findings.NewFinding(
				fmt.Sprintf("Обнаружен защищённый путь: %s (%s)", sp.path, sp.label),
				findings.Info,
				"Information Disclosure",
				"CWE-200",
				u,
				fmt.Sprintf("URL '%s' существует и возвращает HTTP %d "+
					"(требует авторизации).\nРесурс: %s", u, resp.StatusCode, sp.label),
				fmt.Sprintf("Путь %s закрыт авторизацией — это ожидаемое поведение, "+
					"дополнительных действий не требуется. Убедитесь, что 401/403 "+
					"не отдаёт лишних деталей о структуре приложения.", sp.path),
				fmt.Sprintf("curl -sk %s '%s'", curlAuth, u),
			)
			f.Evidence = fmt.Sprintf("HTTP %d\n%s", resp.StatusCode, bodySnippet)
			store.Add(f)
			continue
		}

		ctDisplay := ct
		if ctDisplay == "" {
			ctDisplay = "не указан"
		}
		f := findings.NewFinding(
			fmt.Sprintf("Доступен чувствительный путь: %s (%s)", sp.path, sp.label),
			findings.Medium,
			"Information Disclosure",
			"CWE-200",
			u,
			fmt.Sprintf("URL '%s' возвращает HTTP %d и отличается от "+
				"стандартного fallback-ответа сервера.\n"+
				"Ресурс: %s\n"+
				"Content-Type: %s\n"+
				"Сам по себе нестандартный ответ ещё не доказывает утечку — "+
				"требуется проверка содержимого на признаки реального секрета/исходника.",
				u, resp.StatusCode, sp.label, ctDisplay),
			fmt.Sprintf("1. Закройте доступ к %s на уровне веб-сервера / файрвола.\n"+
				"2. Удалите чувствительные файлы из webroot.\n"+
				"3. Ограничьте Actuator-эндпоинты (management.endpoints.web.exposure.include).", sp.path),
			fmt.Sprintf("curl -sk %s '%s'", curlAuth, u),
		)
		f.Evidence = fmt.Sprintf("HTTP %d\n%s", resp.StatusCode, bodySnippet)

		store.AddCandidate(&findings.Candidate{
			Finding: f,
			Kind:    "disclosure",
			Context: map[string]any{"path": sp.path, "url": u, "label": sp.label},
		})
	}
}

func checkVerboseErrors(session *httpsession.Session, baseURL string, endpoints []crawler.Endpoint, curlAuth string, store *findings.FindingStore) {
	tested := make(map[string]struct{})
	targets := []string{baseURL}
	count := 0
	for _, ep := range endpoints {
		if ep.Method == "GET" {
			targets = append(targets, ep.URL)
			count++
			if count >= 20 {
				break
			}
		}
	}

	for _, url := range targets {
		if _, ok := tested[url]; ok {
			continue
		}
		tested[url] = struct{}{}

		for _, trigger := range errorTriggers[:3] {
			var testURL string
			if !strings.Contains(url, "?") {
				testURL = url + trigger
			} else {
				testURL = url + "&err=" + trigger
			}
			resp, err := session.Request("GET", testURL, httpsession.Options{AllowRedirects: true, Timeout: 8 * time.Second})
			if err != nil {
				continue
			}

			for _, sig := range verboseErrorSignatures {
				loc := sig.FindStringIndex(resp.Body)
				if loc == nil {
					continue
				}
				start := loc[0] - 100
				if start < 0 {
					start = 0
				}
				end := loc[1] + 200
				if end > len(resp.Body) {
					end = len(resp.Body)
				}
				snippet := resp.Body[start:end]

				f := findings.NewFinding(
					"Verbose error messages — утечка стектрейса",
					findings.Medium,
					"Information Disclosure",
					"CWE-209",
					testURL,
					fmt.Sprintf("Приложение возвращает подробные сообщения об ошибках.\n"+
						"Паттерн: %s\n"+
						"Это раскрывает технологический стек, пути файлов и детали реализации.", sig.String()),
					"1. Настройте production error handler: возвращайте только generic сообщение.\n"+
						"2. Логируйте детали ошибки только на сервере.\n"+
						"3. DEBUG=False (Django), displayErrors: false, error_reporting(0) (PHP).",
					fmt.Sprintf("curl -sk %s '%s'", curlAuth, testURL),
				)
				f.Evidence = fmt.Sprintf("...%s...", snippet)
				store.Add(f)
				tested[url] = struct{}{}
				break
			}
			break // one trigger per URL is enough for detection
		}
	}
}

func checkSecretsInResponses(session *httpsession.Session, endpoints []crawler.Endpoint, store *findings.FindingStore) {
	limit := len(endpoints)
	if limit > 30 {
		limit = 30
	}
	for _, ep := range endpoints[:limit] {
		if ep.Method != "GET" && ep.Method != "HEAD" {
			continue
		}
		resp, err := session.Get(ep.URL, nil)
		if err != nil || resp.StatusCode != 200 {
			continue
		}

		for _, sp := range secretPatterns {
			loc := sp.re.FindStringIndex(resp.Body)
			if loc == nil {
				continue
			}
			match := resp.Body[loc[0]:loc[1]]
			f := findings.NewFinding(
				fmt.Sprintf("Секрет в HTTP-ответе: %s", sp.name),
				findings.Critical,
				"Information Disclosure",
				"CWE-312",
				ep.URL,
				fmt.Sprintf("В ответе '%s' обнаружен %s.\n"+
					"Утечка credentials в HTTP-ответах критична для безопасности.", ep.URL, sp.name),
				"1. Немедленно ротируйте скомпрометированные credentials.\n"+
					"2. Не возвращайте секреты в API-ответах.\n"+
					"3. Используйте secret management (Vault, AWS Secrets Manager).\n"+
					"4. Проверьте git-историю на предмет закоммиченных секретов.",
				fmt.Sprintf("curl -sk '%s' | grep -oE '<pattern>'", ep.URL),
			)
			f.Evidence = fmt.Sprintf("Найдено: %s...", truncate(match, 80))
			store.Add(f)
		}
	}
}

var graphqlEndpoints = []string{"/graphql", "/api/graphql", "/graphiql", "/gql", "/query"}

func checkGraphQL(session *httpsession.Session, baseURL, curlAuth string, store *findings.FindingStore) {
	for _, path := range graphqlEndpoints {
		u := strings.TrimRight(baseURL, "/") + path

		probe, err := session.Request("GET", u, httpsession.Options{AllowRedirects: true, Timeout: 8 * time.Second})
		if err != nil {
			continue
		}
		if probe.StatusCode == 404 || probe.StatusCode == 410 {
			continue
		}

		introspection := map[string]any{"query": "{ __schema { types { name fields { name } } } }"}
		resp, err := session.PostJSON(u, introspection, nil)
		if err != nil {
			continue
		}
		if resp.StatusCode == 200 && strings.Contains(resp.Body, "__schema") {
			f := findings.NewFinding(
				"GraphQL: introspection включена (schema disclosure)",
				findings.Medium,
				"Information Disclosure",
				"CWE-200",
				u,
				fmt.Sprintf("GraphQL endpoint '%s' возвращает полную схему через introspection.\n"+
					"Атакующий может восстановить всю структуру API, включая мутации и аргументы.", u),
				"1. Отключите introspection в production:\n"+
					"   GraphQL Shield, depth-limit или параметр disableIntrospection.\n"+
					"2. Если introspection нужна — ограничьте по IP или роли.",
				fmt.Sprintf("curl -sk %s -X POST '%s' \\\n"+
					"  -H 'Content-Type: application/json' \\\n"+
					"  -d '{\"query\":\"{ __schema { types { name } } }\"}", curlAuth, u),
			)
			f.Evidence = "__schema присутствует в ответе на introspection-запрос"
			store.Add(f)
		}

		batchQuery := []map[string]any{introspection, introspection, introspection}
		respBatch, err := session.PostJSON(u, batchQuery, nil)
		if err != nil {
			continue
		}
		if respBatch.StatusCode == 200 {
			var arr []any
			if json.Unmarshal([]byte(respBatch.Body), &arr) == nil {
				f := findings.NewFinding(
					"GraphQL: batch queries включены",
					findings.Low,
					"API Security",
					"CWE-770",
					u,
					"GraphQL endpoint принимает batch-запросы (массив операций в одном HTTP-запросе). "+
						"Это может использоваться для обхода rate limiting или брутфорса.",
					"Ограничьте глубину и количество операций в одном запросе. Отключите batch если не используется.",
					fmt.Sprintf("curl -sk %s -X POST '%s' \\\n"+
						"  -H 'Content-Type: application/json' \\\n"+
						"  -d '[{\"query\":\"{me{id}}\"},{\"query\":\"{me{id}}\"},{\"query\":\"{me{id}}\"}]'", curlAuth, u),
				)
				f.Evidence = fmt.Sprintf("Batch ответ: %s", truncate(respBatch.Body, 200))
				store.Add(f)
			}
		}
	}
}
